// now/build/lib/certs.js
// Ours
const Now = require('../lib')

module.exports = class Certs extends Now {
  ls() {
    return this.retry((bail, attempt) => __async(function*(){
      if (this._debug) {
        console.time(`> [debug] #${attempt} GET now/certs`)
      }

      const res = yield this._fetch('/now/certs')

      if (this._debug) {
        console.timeEnd(`> [debug] #${attempt} GET now/certs`)
      }

      const body = yield res.json()
      return body.certs
    }.call(this)))
  }

  create(cn) {
    return this.createCert(cn)
  }

  renew(cn) {
    return this.createCert(cn, { renew: true })
  }

  put(cn, crt, key, ca) {
    return this.retry((bail, attempt) => __async(function*(){
      if (this._debug) {
        console.time(`> [debug] #${attempt} PUT now/certs`)
      }

      const res = yield this._fetch('/now/certs', {
        method: 'PUT',
        body: {
          domains: [cn],
          ca,
          cert: crt,
          key
        }
      })

      if (this._debug) {
        console.timeEnd(`> [debug] #${attempt} PUT now/certs`)
      }

      if (res.status === 403) {
        return bail(new Error('Unauthorized'))
      }

      const body = yield res.json()

      if (res.status !== 200) {
        if (res.status === 404 || res.status === 400) {
          const err = new Error(body.error.message)
          err.userError = true
          return bail(err)
        }

        throw new Error(body.error.message)
      }

      return body
    }.call(this)))
  }

  delete(cn) {
    return this.retry((bail, attempt) => __async(function*(){
      if (this._debug) {
        console.time(`> [debug] #${attempt} DELETE now/certs/${cn}`)
      }

      const res = yield this._fetch(`/now/certs/${cn}`, { method: 'DELETE' })

      if (this._debug) {
        console.timeEnd(`> [debug] #${attempt} DELETE now/certs/${cn}`)
      }

      if (res.status === 403) {
        return bail(new Error('Unauthorized'))
      }

      const body = yield res.json()

      if (res.status !== 200) {
        if (res.status === 404 || res.status === 400) {
          const err = new Error(body.error.message)
          err.userError = true
          return bail(err)
        }

        throw new Error(body.error.message)
      }

      return body
    }.call(this)))
  }
}

function __async(g){return new Promise(function(s,j){function c(a,x){try{var r=g[x?"throw":"next"](a)}catch(e){j(e);return}r.done?s(r.value):Promise.resolve(r.value).then(c,d)}function d(e){c(e,1)}c()})}
