// now/build/lib/domains.js
// Native
const { encode: encodeQuery } = require('querystring')

// Packages
const chalk = require('chalk')

// Ours
const Now = require('../lib')
const cfg = require('../lib/cfg')
const isZeitWorld = require('./is-zeit-world')
const { DNS_VERIFICATION_ERROR } = require('./errors')

const domainRegex = /^((?=[a-z0-9-]{1,63}\.)(xn--)?[a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,63}$/

module.exports = class Domains extends Now {
  ls() {return __async(function*(){
    return this.listDomains()
  }.call(this))}

  rm(name) {return __async(function*(){
    return this.retry((bail, attempt) => __async(function*(){
      if (this._debug) {
        console.time(`> [debug] #${attempt} DELETE /domains/${name}`)
      }

      const res = yield this._fetch(`/domains/${name}`, { method: 'DELETE' })

      if (this._debug) {
        console.timeEnd(`> [debug] #${attempt} DELETE /domains/${name}`)
      }

      if (res.status === 403) {
        return bail(new Error('Unauthorized'))
      }

      if (res.status !== 200) {
        const body = yield res.json()
        throw new Error(body.error.message)
      }
    }.call(this)))
  }.call(this))}

  add(domain, skipVerification, isExternal) {return __async(function*(){
    if (!domainRegex.test(domain)) {
      const err = new Error(
        `The supplied value ${chalk.bold(`"${domain}"`)} is not a valid domain.`
      )
      err.userError = true
      throw err
    }

    if (skipVerification || isExternal) {
      return this.setupDomain(domain, { isExternal })
    }

    let ns

    try {
      console.log('> Verifying nameservers…')
      const res = yield this.getNameservers(domain)
      ns = res.nameservers
    } catch (err) {
      const err2 = new Error(
        `Unable to fetch nameservers for ${chalk.underline(chalk.bold(domain))}.`
      )
      err2.userError = true
      throw err2
    }

    if (isZeitWorld(ns)) {
      console.log(`> Verification ${chalk.bold('OK')}!`)
      return this.setupDomain(domain)
    }

    if (this._debug) {
      console.log(
        `> [debug] Supplied domain "${domain}" has non-zeit nameservers`
      )
    }

    const err3 = new Error(DNS_VERIFICATION_ERROR)
    err3.userError = true
    throw err3
  }.call(this))}

  status(name) {return __async(function*(){
    if (!name) {
      throw new Error('`domain` is not defined')
    }

    const query = encodeQuery({ name })

    return this.retry((bail, attempt) => __async(function*(){
      if (this._debug) {
        console.time(`> [debug] #${attempt} GET /domains/status?${query}`)
      }

      const res = yield this._fetch(`/domains/status?${query}`)
      const json = yield res.json()

      if (this._debug) {
        console.timeEnd(`> [debug] #${attempt} GET /domains/status?${query}`)
      }

      return json.available
    }.call(this)))
  }.call(this))}

  price(name) {return __async(function*(){
    if (!name) {
      throw new Error('`domain` is not defined')
    }

    const query = encodeQuery({ name })

    return this.retry((bail, attempt) => __async(function*(){
      if (this._debug) {
        console.time(`> [debug] #${attempt} GET /domains/price?${query}`)
      }

      const res = yield this._fetch(`/domains/price?${query}`)
      const json = yield res.json()

      if (this._debug) {
        console.timeEnd(`> [debug] #${attempt} GET /domains/price?${query}`)
      }

      return json.price
    }.call(this)))
  }.call(this))}

  buy(name) {return __async(function*(){
    const { token } = yield cfg.read()
    if (!name) {
      throw new Error('`name` is not defined')
    }

    const rawBody = { name }

    if (name.startsWith('test')) {
      rawBody.dev = true
    }

    const body = JSON.stringify(rawBody)

    return this.retry((bail, attempt) => __async(function*(){
      if (this._debug) {
        console.time(`> [debug] #${attempt} GET /domains/buy`)
      }
      const res = yield this._fetch(`/domains/buy`, {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${token}`
        },
        body
      })
      const json = yield res.json()

      if (this._debug) {
        console.timeEnd(`> [debug] #${attempt} GET /domains/buy`)
      }

      if ([400, 403, 500, 503].includes(res.status)) {
        const e = new Error()
        e.code = json.error.code
        e.message = json.error.message
        return bail(e)
      }

      return json
    }.call(this)))
  }.call(this))}
}

function __async(g){return new Promise(function(s,j){function c(a,x){try{var r=g[x?"throw":"next"](a)}catch(e){j(e);return}r.done?s(r.value):Promise.resolve(r.value).then(c,d)}function d(e){c(e,1)}c()})}
