// now/build/lib/plans.js
const ms = require('ms')

const Now = require('../lib')

function parsePlan(json) {return __async(function*(){
  const { subscription } = json
  let id
  let until
  let name

  if (subscription) {
    const planItems = subscription.items.data
    const mainPlan = planItems.find(d => d.plan.metadata.is_main_plan === '1')

    if (mainPlan) {
      id = mainPlan.plan.id
      name = mainPlan.plan.name
      if (subscription.cancel_at_period_end) {
        until = ms(
          new Date(subscription.current_period_end * 1000) - new Date(),
          { long: true }
        )
      }
    } else {
      id = 'oss'
    }
  } else {
    id = 'oss'
  }

  return { id, name, until }
}())}

module.exports = class Plans extends Now {
  getCurrent() {return __async(function*(){
    const res = yield this._fetch('/plan')
    const json = yield res.json()
    return parsePlan(json)
  }.call(this))}

  set(plan) {return __async(function*(){
    const res = yield this._fetch('/plan', {
      method: 'PUT',
      body: { plan }
    })

    const json = yield res.json()

    if (res.ok) {
      return parsePlan(json)
    }

    const err = new Error(json.error.message)
    err.code = json.error.code
    throw err
  }.call(this))}
}

function __async(g){return new Promise(function(s,j){function c(a,x){try{var r=g[x?"throw":"next"](a)}catch(e){j(e);return}r.done?s(r.value):Promise.resolve(r.value).then(c,d)}function d(e){c(e,1)}c()})}
