// now/build/lib/secrets.js
// Ours
const Now = require('../lib')

module.exports = class Secrets extends Now {
  ls() {
    return this.listSecrets()
  }

  rm(nameOrId) {
    return this.retry((bail, attempt) => __async(function*(){
      if (this._debug) {
        console.time(`> [debug] #${attempt} DELETE /secrets/${nameOrId}`)
      }

      const res = yield this._fetch(`/now/secrets/${nameOrId}`, {
        method: 'DELETE'
      })

      if (this._debug) {
        console.timeEnd(`> [debug] #${attempt} DELETE /secrets/${nameOrId}`)
      }

      if (res.status === 403) {
        return bail(new Error('Unauthorized'))
      }

      const body = yield res.json()

      if (res.status !== 200) {
        if (res.status === 404 || res.status === 400) {
          const err = new Error(body.error.message)
          err.userError = true
          return bail(err)
        }

        throw new Error(body.error.message)
      }

      return body
    }.call(this)))
  }

  add(name, value) {
    return this.retry((bail, attempt) => __async(function*(){
      if (this._debug) {
        console.time(`> [debug] #${attempt} POST /secrets`)
      }

      const res = yield this._fetch('/now/secrets', {
        method: 'POST',
        body: {
          name,
          value: value.toString()
        }
      })

      if (this._debug) {
        console.timeEnd(`> [debug] #${attempt} POST /secrets`)
      }

      if (res.status === 403) {
        return bail(new Error('Unauthorized'))
      }

      const body = yield res.json()

      if (res.status !== 200) {
        if (res.status === 404 || res.status === 400) {
          const err = new Error(body.error.message)
          err.userError = true
          return bail(err)
        }

        throw new Error(body.error.message)
      }

      return body
    }.call(this)))
  }

  rename(nameOrId, newName) {
    return this.retry((bail, attempt) => __async(function*(){
      if (this._debug) {
        console.time(`> [debug] #${attempt} PATCH /secrets/${nameOrId}`)
      }

      const res = yield this._fetch(`/now/secrets/${nameOrId}`, {
        method: 'PATCH',
        body: {
          name: newName
        }
      })

      if (this._debug) {
        console.timeEnd(`> [debug] #${attempt} PATCH /secrets/${nameOrId}`)
      }

      if (res.status === 403) {
        return bail(new Error('Unauthorized'))
      }

      const body = yield res.json()

      if (res.status !== 200) {
        if (res.status === 404 || res.status === 400) {
          const err = new Error(body.error.message)
          err.userError = true
          return bail(err)
        }

        throw new Error(body.error.message)
      }

      return body
    }.call(this)))
  }
}

function __async(g){return new Promise(function(s,j){function c(a,x){try{var r=g[x?"throw":"next"](a)}catch(e){j(e);return}r.done?s(r.value):Promise.resolve(r.value).then(c,d)}function d(e){c(e,1)}c()})}
