// now/build/lib/user.js
const _fetch = require('node-fetch')

function _filter(data) {
  data = data.user

  return {
    userId: data.uid,
    username: data.username,
    email: data.email
  }
}

/**
 * Gets all the info we have about an user
 *
 * @param  {Object} fetch    Optionally, _our_ `fetch` can be passed here
 * @param  {String} token    Only necessary if `fetch` is undefined
 * @param  {String} apiUrl   Only necessary if `fetch` is undefined
 * @param  {Boolean} filter  If `true`, the `filter` used will to the data
 *                           before returning
 * @return {Object}
 */
function get(
  { fetch, token, apiUrl = 'https://api.zeit.co', filter = true } = {}
) {return __async(function*(){
  let headers = {}
  const endpoint = '/www/user'
  const url = fetch ? endpoint : apiUrl + endpoint

  if (!fetch) {
    headers = {
      Authorization: `Bearer ${token}`
    }
    fetch = _fetch
  }

  try {
    const res = yield fetch(url, { headers })

    const json = yield res.json()

    if (filter) {
      return _filter(json)
    }
    return json
  } catch (err) {
    return {}
  }
}())}

module.exports = {
  get,
  filter: _filter
}

function __async(g){return new Promise(function(s,j){function c(a,x){try{var r=g[x?"throw":"next"](a)}catch(e){j(e);return}r.done?s(r.value):Promise.resolve(r.value).then(c,d)}function d(e){c(e,1)}c()})}
